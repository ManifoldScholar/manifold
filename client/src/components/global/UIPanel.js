import React, { Component } from "react";
import PropTypes from "prop-types";
import classNames from "classnames";
import FocusTrap from "focus-trap-react";

export default class UIPanel extends Component {
  static propTypes = {
    id: PropTypes.string,
    visibility: PropTypes.object,
    bodyComponent: PropTypes.func,
    hidePanel: PropTypes.func
  };

  render() {
    const visibility = this.props.visibility[this.props.id];
    const visibilityClass = classNames({
      "panel-hidden": !visibility,
      "panel-visible": visibility
    });

    return (
      <div>
        {visibility ? (
          <div className={visibilityClass}>
            <FocusTrap
              focusTrapOptions={{
                clickOutsideDeactivates: true,
                escapeDeactivates: false
              }}
            >
              <button
                className="screen-reader-text"
                onClick={this.props.hidePanel}
              >
                Close Panel
              </button>
              {/* Second argument as props */}
              {React.createElement(this.props.bodyComponent, { ...this.props })}
            </FocusTrap>
          </div>
        ) : null}
      </div>
    );
  }
}
