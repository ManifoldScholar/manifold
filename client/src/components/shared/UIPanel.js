import React, { Component, PropTypes } from 'react';
import classNames from 'classnames';

export default class UIPanel extends Component {
  static propTypes = {
    id: PropTypes.string,
    visibility: PropTypes.object,
    bodyComponent: PropTypes.func
  };

  render = () => {
    const visibilityClass = classNames({
      'panel-hidden': !this.props.visibility[this.props.id],
      'panel-visible': this.props.visibility[this.props.id]
    });
    return (
        <div className={visibilityClass}>
          {/* Second argument as props */}
          {React.createElement(this.props.bodyComponent, {...this.props})}
        </div>
    );
  };
}

