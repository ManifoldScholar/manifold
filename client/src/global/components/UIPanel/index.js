import React from "react";
import PropTypes from "prop-types";
import classNames from "classnames";
import { CSSTransition } from "react-transition-group";
import FocusTrap from "focus-trap-react";

export default function UIPanel(props) {
  const visibility = props.visibility[props.id];
  const visibilityClass = classNames({
    "panel-hidden": !visibility,
    "panel-visible": visibility
  });

  const handleOutsideClick = e => {
    const header = document.querySelector(".reader-header");
    if (header?.contains(e.target)) {
      return true;
    }
    return props.hidePanel(e);
  };

  return (
    <CSSTransition
      key={props.id}
      in={visibility}
      classNames="panel"
      timeout={200}
      unmountOnExit
    >
      <div className={visibilityClass}>
        <FocusTrap
          focusTrapOptions={{
            allowOutsideClick: handleOutsideClick,
            escapeDeactivates: e => props.hidePanel(e)
          }}
        >
          <div>
            {React.createElement(props.bodyComponent, {
              ...props,
              closeCallback: props.hidePanel
            })}
          </div>
        </FocusTrap>
      </div>
    </CSSTransition>
  );
}

UIPanel.displayName = "Global.UIPanel";

UIPanel.propTypes = {
  id: PropTypes.string,
  visibility: PropTypes.object,
  bodyComponent: PropTypes.elementType,
  hidePanel: PropTypes.func
};
