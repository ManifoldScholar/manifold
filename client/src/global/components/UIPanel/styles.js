import styled from "@emotion/styled";
import { defaultTransitionProps } from "theme/styles/mixins";

export const Panel = styled.div`
  .panel {
    transition: transform ${defaultTransitionProps},
      opacity ${defaultTransitionProps};
  }

  &[inert] .panel {
    opacity: var(--Panel-starting-opacity);
    transform: var(--Panel-starting-transform);
  }
`;
