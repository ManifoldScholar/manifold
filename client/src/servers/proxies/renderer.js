import config from "config";
import proxy from "http-proxy-middleware";
import ch from "../../helpers/consoleHelpers";
import createStore from "../../store/createStore";
import exceptionRenderer from "../../helpers/exceptionRenderer";
import manifoldBootstrap from "bootstrap";
import CookieHelper from "helpers/cookie/Server";

const ssrRenderUrl = `http://${config.services.client.domain}:${config.services.client.sparePort}`;

export default function makeRendererProxy(requestHandler) {
  const reactServerProxy = proxy({
    target: ssrRenderUrl,
    changeOrigin: true,
    logLevel: "silent",

    onError: (err, req, res) => {
      ch.error(`Error proxying to port ${ssrRenderUrl}`);
      ch.error(`Perhaps the SSR service is in the process of reloading?`);
      ch.error(`Falling back to SSR rescue service only`);

      const store = createStore();

      const render = () => {
        try {
          requestHandler(req, res);
        } catch (error) {
          if (error.code === "MODULE_NOT_FOUND") {
            const msg =
              "Waiting for initial Webpack build to complete. Wait a few seconds " +
              "and reload.";
            ch.error(msg);
            res.setHeader("Content-Type", "text/html");
            res.end(msg);
          } else {
            ch.error(
              `Rendering fallback failed to render inn server-developmet.js`
            );
            res.setHeader("Content-Type", "text/html");
            res.end(exceptionRenderer(error));
          }
        }
      };

      const promises = [];
      const cookie = new CookieHelper(req, res);
      promises.push(manifoldBootstrap(store.getState, store.dispatch, cookie));
      Promise.all(promises).then(render, render);
    }
  });
  return reactServerProxy;
}
