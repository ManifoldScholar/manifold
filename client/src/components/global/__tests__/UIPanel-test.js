import React from 'react';
import renderer from 'react-test-renderer';
import UIPanel from '../UIPanel';

describe("Global.UIPanel component", () => {

  function bodyComponentMock() {
    return (
      <div>
        Render me
      </div>
    );
  }

  it('renders correctly', () => {
    const component = renderer.create(
      <UIPanel
        id="show"
        visibility={{
          show: true
        }}
        bodyComponent={bodyComponentMock}
      />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

