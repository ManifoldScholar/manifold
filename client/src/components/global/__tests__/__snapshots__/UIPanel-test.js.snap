// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Global.UIPanel component renders correctly 1`] = `
<div
  className="panel-visible"
>
  <focus-trap-react
    active={true}
    focusTrapOptions={
      Object {
        "clickOutsideDeactivates": true,
        "escapeDeactivates": false,
      }
    }
  >
    <button
      className="screen-reader-text"
      onClick={undefined}
    >
      Close Panel
    </button>
    <div>
      Render me
    </div>
  </focus-trap-react>
</div>
`;
